consensus/src/block_preparer.rs

```diff
@@ -36,16 +37,31 @@ impl BlockPreparer {
         }
     }
 
-    pub async fn prepare_block(&self, block: &Block) -> ExecutorResult<Vec<SignedTransaction>> {
+    pub async fn prepare_block(
+        &self,
+        block: &Block,
+        block_qc_fut: Shared<impl Future<Output = Option<Arc<QuorumCert>>>>,
+    ) -> ExecutorResult<Vec<SignedTransaction>> {
         fail_point!("consensus::prepare_block", |_| {
             use aptos_executor_types::ExecutorError;
             use std::{thread, time::Duration};
             thread::sleep(Duration::from_millis(10));
             Err(ExecutorError::CouldNotGetData)
         });
         let start_time = Instant::now();
-        let (txns, max_txns_from_block_to_execute) =
-            self.payload_manager.get_transactions(block).await?;
+
+        let mut block_voters = None;
+        let (txns, max_txns_from_block_to_execute) = loop {
```

Comment:
I'm not sure how this works? it reads like it always get_transactions with None. 

it creates two initial futures first (which always uses None) then select on the result and qc, regardless when the result comes back, it returns so it'll never use the block_voters to get_transactions again?

Reply:
Line 58 sets `block_voters` to Some() and continuing the next iteration of loop? If the block_qc_fut returns some QC, then we will extract the voters bitvec and use it for the next iteration. We only break if get_transactions returns a result.

Reply:
I figured what confused me is the new iteration of the loop, maybe it's less confusing to write it like this?
```
select! {
  qc => {
    get_transaction()?
  }
  txns = get_transactions => txns
}
```

Reply:
I see, yeah. Updated the code.